### DISCUSSION TITLE
Multi-line forward reference type annotations

### DISCUSSION THREAD

**Discussion by smheidrich** [microsoft/pyright#4379] [Q&A] [answered]
It seems to me like Pyright doesn't let you break up a type annotated as a [forward reference](https://peps.python.org/pep-0484/#forward-references) (i.e. a string literal) over multiple lines in any way. Consider this minimal example:

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class SomeSuperLongTypeThatWillMakeUsWantToBreakALine(Generic[T]):
    pass

# splitting a regular annotation over multiple lines works fine:
a: SomeSuperLongTypeThatWillMakeUsWantToBreakALine[
    SomeSuperLongTypeThatWillMakeUsWantToBreakALine
]

# "forward reference" annotations work fine:
b: "SomeSuperLongTypeThatWillMakeUsWantToBreakALine"

# but not both together:
c: (
    "SomeSuperLongTypeThatWillMakeUsWantToBreakALine["
    "SomeSuperLongTypeThatWillMakeUsWantToBreakALine]"
)

# also not if you use a single string literal:
d: (
    "SomeSuperLongTypeThatWillMakeUsWantToBreakALine[\
    SomeSuperLongTypeThatWillMakeUsWantToBreakALine]"
)
```

For `c`'s annotation, Pyright's complaint is pretty straightforward:

```
error: Type annotations cannot span multiple string literals
error: Expected type but received a string literal (reportGeneralTypeIssues)
```

Mypy allows this, so I'm wondering if that's really something required by a PEP and even if so, if it wouldn't be better to be more lenient than allowed by the standard here...

`d`'s annotation is an attempt at an uglier workaround, because suppressing the line break with a trailing `\` should make both lines count as a single string literal, but alas, here too Pyright complains:

```
error: Type annotations cannot contain escape characters
error: "[" was not closed
error: "SomeSuperLongTypeThatWillMakeUsWantToBreakALin" is not defined (reportUndefinedVariable)
```

(This is besides the point of this question, but the last error is a bit puzzling to me: Did the mere presence of the escape character in the previous line somehow mess up Pyright's parsing of subsequent type names?)

So with both of these options shot down, is there another way to have a forward reference style annotation span multiple lines?

**Comment by erictraut** [comment, marked as answer]
See [this issue](https://github.com/microsoft/pyright/issues/4359) for details.

**Reply by smheidrich** [reply]
Thanks!
